pkg/datapath/tables/node_address.go

```diff
@@ -0,0 +1,365 @@
+	// If --nodeport-addresses is set, then all addresses on native
+	// devices that are contained within the specified CIDRs are chosen.
+	// If it is not set, then only the primary IPv4 and/or IPv6 address
+	// of each native device is used.
+	NodePort bool
+
+	// Primary is true if this is the primary IPv4 or IPv6 address of this device.
+	// This is mainly used to pick the address for BPF masquerading.
+	Primary bool
+
+	// DeviceName is the name of the network device from which this address
+	// is derived from.
+	DeviceName string
+}
+
+func (n *NodeAddress) IP() net.IP {
+	return n.Addr.AsSlice()
+}
+
+func (n *NodeAddress) String() string {
+	return fmt.Sprintf("%s (%s)", n.Addr, n.DeviceName)
+}
+
+type NodeAddressConfig struct {
+	NodePortAddresses []*cidr.CIDR `mapstructure:"nodeport-addresses"`
+}
+
+var (
+	// NodeAddressIndex is the primary index for node addresses:
+	//
+	//   var nodeAddresses Table[NodeAddress]
+	//   nodeAddresses.First(txn, NodeAddressIndex.Query(netip.MustParseAddr("1.2.3.4")))
+	NodeAddressIndex = statedb.Index[NodeAddress, netip.Addr]{
+		Name: "id",
+		FromObject: func(a NodeAddress) index.KeySet {
+			return index.NewKeySet(index.NetIPAddr(a.Addr))
+		},
+		FromKey: func(addr netip.Addr) []byte {
+			return index.NetIPAddr(addr)
+		},
+		Unique: true,
+	}
+
+	NodeAddressDeviceNameIndex = statedb.Index[NodeAddress, string]{
+		Name: "name",
+		FromObject: func(a NodeAddress) index.KeySet {
+			return index.NewKeySet(index.String(a.DeviceName))
+		},
+		FromKey: index.String,
+		Unique:  false,
+	}
+
+	NodeAddressTableName statedb.TableName = "node-addresses"
+
+	// NodeAddressCell provides Table[NodeAddress] and a background controller
+	// that derives the node addresses from the low-level Table[*Device].
+	//
+	// The Table[NodeAddress] contains the actual assigned addresses on the node,
+	// but not for example external Kubernetes node addresses that may be merely
+	// NATd to a private address. Those can be queried through [node.LocalNodeStore].
+	NodeAddressCell = cell.Module(
+		"node-address",
+		"Table of node addresses derived from system network devices",
+
+		statedb.NewPrivateRWTableCell[NodeAddress](NodeAddressTableName, NodeAddressIndex, NodeAddressDeviceNameIndex),
+		cell.Provide(
+			newNodeAddressTable,
+			newAddressScopeMax,
+		),
+		cell.Config(NodeAddressConfig{}),
+	)
+
+	// NodeAddressTestTableCell provides Table[NodeAddress] and RWTable[NodeAddress]
+	// for use in tests of modules that depend on node addresses.
+	NodeAddressTestTableCell = statedb.NewTableCell[NodeAddress](
+		NodeAddressTableName,
+		NodeAddressIndex,
+	)
+)
+
+const (
+	nodeAddressControllerMinInterval = 100 * time.Millisecond
+)
+
+type AddressScopeMax uint8
+
+func newAddressScopeMax(cfg NodeAddressConfig, daemonCfg *option.DaemonConfig) (AddressScopeMax, error) {
+	return AddressScopeMax(daemonCfg.AddressScopeMax), nil
+}
+
+func (cfg NodeAddressConfig) getNets() []*net.IPNet {
```

Comment:
Does this have to be old-style `net.IPNet` or can it be something from `netip`? Given this is new code, I'm hoping for the latter.

Reply:
It's `net.IPNet` for `ip.NetsContainsAny`. Could add a `netip.Addr` variant for `NetsContainsAny` perhaps? Or do you have other ideas?